### DISCUSSION TITLE
Can't save Refs in pinia store ?

### DISCUSSION THREAD

**Discussion by Chlac** [nuxt/nuxt#19142] [Questions] [answered]
Hi all,

I'm very confused about a Typescript error I'm getting.

I'm trying to fetch some data with **useFetch in the pinia store and save the obtained Ref** . But **it somehow always gets unwrapped to the referenced value without me using the ".value"**, and I absolutely can't save the whole Ref ?.... Here's the code :

```
export const useBoxsStore = defineStore({
  id: "boxs",

  state: () => {
    return {
      boxs: ref(null) as Ref<any | null>,
    };
  },

  actions: {
    async initBoxs(): Promise<void> {
      const API_URL = useRuntimeConfig().public.apiURL;

      const { data, pending, error, refresh } = await useFetch<Box>(
        API_URL + "/boxs/get.php",
        {
          method: "GET",
        }
      );

      const test: Ref<Box | null> = data;
      this.boxs = data;


      console.log(test.value); // This works
      console.log(this.boxs); // This works and prints the same thing
      console.log(this.boxs.value); // This doesn't work / undefined
    },
});
```

**If I put my mouse over any reference to "this.boxs" I get the tooltip : "(property) boxs: any" instead of "Ref<any | null".
I've tried many different variations of this, with interfaces and whatnot. Ref always gets unpacked somehow.**

1. Is it impossible to store the Ref type outside its initial declaration scope (not sure what I even mean) ?
2. Is it because you shouldn't use a Ref type anywhere outside "setup" or something ? I wasn't even planning on calling this action from a setup function, but from a plugin, so that would be a bit absurd, but............
3. Is there a different way to achieve what I'm trying to do ? Basically I wanted to store the Ref and the refresh function, so I can call it whenever, from different components or stores. I also want the fetching thingy inside the store to be called just once, because different pages / components might need that same data, but I see in the docs that useFetch has some sort of "de-duplicate" function, if the url is the same. Does that mean, if a different page or component makes the same call, it would use some cached data instead of re-fetching it ?..... Does this mean I should do any data fetching in the setup functions, not in the store, and not worry about duplicate calls ?...

Sorry for the multi-part question, but I find the docs and lack of use-cases / examples very confusing...

Thx for any tips.

**Comment by TheAlexLichter** [comment, marked as answer]
Have you tried Pinia's [`setupStore` declaration](https://pinia.vuejs.org/core-concepts/#setup-stores)? That should be closer to what you want.

**Reply by Chlac** [reply]
Thank you so much, this saves so much hassle indeed. It works !

**Comment by AloisSeckar** [comment]
Pinia's `state` itself is a `reactive`, which means it provides reactivity of its contents out of the box. What is the exact usecase and benefit of trying to store another ref inside? Sorry if I miss something obvious there...

**Reply by Chlac** [reply]
It's probably me that's missing something, but I just thought : If I call the "refresh" method returned by "useFetch" while missing the original Ref, it won't get updated ? But maybe that's absolutely not how you use the refresh method ?

**Reply by AloisSeckar** [reply]
Thats the "obvious" I was missing - so far I didn't need to use refresh on my fetched data, so I didnt take that to account. Sorry for misleading comment :(
